text/0000-variadic.md

```diff
@@ -0,0 +1,250 @@
+ownership of the object. To continue using the object after a call to these
+functions, pass a clone of it instead.
+
+Conversely, an `unsafe extern "C"` function written in Rust may accept a
+`VaList` parameter, to allow implementing the `v` variants of such functions in
+Rust. Such a function must not specify the lifetime.
+
+Defining a variadic function, or calling any of these new functions, requires a
+feature-gate, `c_variadic`.
+
+Sample Rust code exposing a variadic function:
+
+```rust
+#![feature(c_variadic)]
+
+#[no_mangle]
+pub unsafe extern "C" fn func(fixed: u32, mut args: ...) {
+    let x: u8 = args.arg();
+    let y: u16 = args.arg();
+    let z: u32 = args.arg();
+    println!("{} {} {} {}", fixed, x, y, z);
+}
+```
+
+Sample C code calling that function:
+
+```c
+#include <stdint.h>
+
+void func(uint32_t fixed, ...);
+
+int main(void)
+{
+    uint8_t x = 10;
+    uint16_t y = 15;
+    uint32_t z = 20;
+    func(5, x, y, z);
+    return 0;
+}
+```
+
+Compiling and linking these two together will produce a program that prints:
+
+```text
+5 10 15 20
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+LLVM already provides a set of intrinsics, implementing `va_start`, `va_arg`,
+`va_end`, and `va_copy`. The compiler will insert a call to the `va_start`
+intrinsic at the start of the function to provide the `VaList` argument (if
+used). The implementation of `VaList::arg` will call `va_arg`.  The
+implementation of `Clone` for `VaList` wil call `va_copy`. The compiler will
+ensure that a call to `va_end` occurs exactly once on every `VaList` at an
+appropriate time, similar to drop semantics. (This may occur via an
+implementation of `Drop`, but this must take into account the semantics of
+passing a `VaList` to or from an `extern "C"` function.)
+
+`VaList` may become a language item (`#[lang="VaList"]`) to attach the
+appropriate compiler handling.
+
+The compiler may need to handle the type `VaList` specially, in order to
+provide the desired drop semantics at FFI boundaries. In particular, some
+platforms pass `va_list` by value, and others pass it by pointer; the standard
+leaves unspecified whether changes made in the called function appear in the
+caller's copy. Rust must match the underlying C semantics, to allow passing
+VaList values between C and Rust. To avoid forcing variadic functions to cope
+with these platform-specific differences, the compiler should ensure that the
+type behaves as if it has `Drop` semantics, and has `va_end` called on the
+original `VaList` as well as any clones of it. For instance, passing a `VaList`
+to a `vprintf` function as declared in this RFC semantically passes ownership
+of that `VaList`, and prevents calling the `arg` function again in the caller,
+but it remains the caller's responsibility to call `va_end`, so the compiler
+needs to insert the appropriate drop glue at the FFI boundary.
```

Comment:
I'm not entirely sure what this means, but it's worth noting that if `va_list` is passed to an object, it should not be dropped to allow for implementing an API like `vprintf`. For example, in following example.

```rust
pub unsafe extern "C" fn vprintf(format: *const c_char, args: VaList) {
    let args_clone = args.clone();
}
```

`args` should not be dropped (otherwise it's not possible to implement API like `vprintf` in Rust), while `args_clone` should be dropped. This maps the behaviour to the following C program.

```c
void vprintf(const char *format, va_list args) {
    va_list args_clone;
    va_copy(args_clone, args);
    va_end(args_clone);
}
```

Of course, this matches C semantics, so it may be different for LLVM.

Reply:
I've added some text to clarify this, thank you.

I've also added an explanation of relevant C standards requirements.